allensdk/brain_observatory/filtered_running_speed/__main__.py

```diff
@@ -0,0 +1,954 @@
+
+    # Get nan indices to propogate to the cumulative sum (otherwise
+    # treated as 0)
+    unwrapped_nans = np.array(np.isnan(unwrapped_diff)).nonzero()
+    summed_diff = np.nancumsum(unwrapped_diff) + vsig[0]    # Add the baseline
+
+    summed_diff[unwrapped_nans] = np.nan
+
+    return summed_diff
+
+
+def _shift(
+        arr: Iterable,
+        periods: int = 1,
+        fill_value: float = np.nan) -> np.ndarray:
+    """
+    Shift index of an iterable (array-like) by desired number of
+    periods with an optional fill value (default = NaN).
+
+    Parameters
+    ----------
+    arr: Iterable (array-like)
+        Iterable containing numeric data. If int, will be converted to
+        float in returned object.
+    periods: int (default=1)
+        The number of elements to shift.
+    fill_value: float (default=np.nan)
+        The value to fill at the beginning of the shifted array
+    Returns
+    -------
+    np.ndarray (1d)
+        Copy of input object as a 1d array, shifted.
+    """
+    if periods <= 0:
+        raise ValueError("Can only shift for periods > 0.")
+
+    if fill_value is None:
+        fill_value = np.nan
+
+    if isinstance(fill_value, float):
+        # Circumvent issue if int-like array with np.nan as fill
+        shifted = np.roll(arr, periods).astype(float)
+
+    else:
+        shifted = np.roll(arr, periods)
+
+    shifted[:periods] = fill_value
+
+    return shifted
+
+
+def _identify_wraps(vsig: Iterable, *,
+                    min_threshold: float = 1.5,
+                    max_threshold: float = 3.5):
+    """
+    Identify "wraps" in the voltage signal. In practice, this is when
+    the encoder voltage signal crosses 5V and wraps to 0V, or
+    vice-versa.
+
+    Argument defaults and implementation suggestion via @dougo
+
+    Parameters
+    ----------
+    vsig: Iterable (array-like)
+        1d array-like iterable of voltage signal
+    min_threshold: float (default=1.5)
+        The min_threshold value that must be crossed to be considered
+        a possible wrapping point.
+    max_threshold: float (default=3.5)
+        The max threshold value that must be crossed to be considered
+        a possible wrapping point.
+
+    Returns
+    -------
+    Tuple
+        Tuple of ([indices of positive wraps], [indices of negative wraps])
+    """
+    # Compare against previous value
+    shifted_vsig = _shift(vsig)
+
+    if not isinstance(vsig, np.ndarray):
+        vsig = np.array(vsig)
+
+    # Suppress warnings for when comparing to nan values
+    with np.errstate(invalid='ignore'):
+        pos_wraps = np.asarray(
+            np.logical_and(vsig < min_threshold, shifted_vsig > max_threshold)
+            ).nonzero()[0]
+        neg_wraps = np.asarray(
+            np.logical_and(vsig > max_threshold, shifted_vsig < min_threshold)
+            ).nonzero()[0]
+
+    return pos_wraps, neg_wraps
+
+
+def extract_running_speeds(
```

Comment:
Has same name as [here](https://github.com/AllenInstitute/AllenSDK/blob/master/allensdk/brain_observatory/extract_running_speed/__main__.py#L50). What is the relationship between the 2? Able to combine the 2 somehow?

Reply:
This has been refactored.